sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala

```diff
@@ -170,14 +170,57 @@ class PlannerSuite extends SparkFunSuite with SQLTestUtils {
 
       // Disable broadcast join
       withSQLConf(SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "-1") {
+        val joins = Array("JOIN", "LEFT OUTER JOIN", "RIGHT OUTER JOIN", "FULL OUTER JOIN")
+        var i = 0
+        while (i < joins.length) {
+          var j = 0
+          while (j < joins.length) {
+            val firstJoin: String = joins(i)
+            val secondJoin: String = joins(j)
+
+            {
+              val numExchanges: Int = sql(
+                s"""
+                |SELECT *
+                |FROM
+                |  normal $firstJoin small ON (normal.key = small.key)
+                |  $secondJoin tiny ON (small.key = tiny.key)
+              """.stripMargin
+              ).queryExecution.executedPlan.collect {
+                case exchange: Exchange => exchange
+              }.length
+              assert(numExchanges === 3)
+            }
+
+            {
+              val numExchanges: Int = sql(
+                s"""
+                |SELECT *
+                |FROM
+                |  normal $firstJoin small ON (normal.key = small.key)
+                |  $secondJoin tiny ON (normal.key = tiny.key)
+              """.stripMargin
+              ).queryExecution.executedPlan.collect {
+                case exchange: Exchange => exchange
+              }.length
+              assert(numExchanges === 3)
+            }
+
+            j += 1
+          }
+          i += 1
+        }
+
         {
-          val numExchanges = sql(
-            """
-              |SELECT *
-              |FROM
-              |  normal JOIN small ON (normal.key = small.key)
-              |  JOIN tiny ON (small.key = tiny.key)
-            """.stripMargin
+          val numExchanges: Int = sql(
+            s"""
+                |SELECT small.key, count(*)
+                |FROM
+                |  normal JOIN small ON (normal.key = small.key)
+                |  JOIN tiny ON (small.key = tiny.key)
+                |GROUP BY
+                |  small.key
+              """.stripMargin
```

Comment:
And I have some unit test at https://github.com/apache/spark/pull/6413/files#diff-f5a576876a5505c7681e5c8506d07250R33

Since we support the `PartitioningCollection` now, we probably need to add more unit test for the key from the other join side.